build_runner/test/server/serve_integration_test.dart

```diff
@@ -108,6 +109,74 @@ example:file://fake/pkg/path
     final response = await handler(new Request('GET', getNew));
     expect(await response.readAsString(), 'NEW');
   });
+
+  group(r'/$graph', () {
+    FutureOr<Response> getResponse(String path) =>
+        handler(new Request('GET', Uri.parse('http://localhost/\$graph$path')));
+
+    for (var slashOrNot in ['', '/']) {
+      test('/\$graph$slashOrNot should (try to) send the HTML page', () async {
+        try {
+          await getResponse(slashOrNot);
+          fail('Assets are not wired up. Expecting this to throw.');
+        } catch (e) {
+          expect(e, new isInstanceOf<AssetNotFoundException>());
+          expect((e as AssetNotFoundException).assetId,
+              new AssetId.parse('build_runner|lib/src/server/graph_viz.html'));
+        }
+      });
+    }
+
+    void test404(String testName, String path, String expected) {
+      test(testName, () async {
+        var response = await getResponse(path);
+
+        expect(response.statusCode, 404);
+        expect(await response.readAsString(), expected);
+      });
+    }
+
+    test404('404s on an unsupported URL', '/bob', 'Bad request: "bob".');
+    test404('404s on an unsupported URL', '/bob/?q=bob',
+        'Bad request: "bob/?q=bob".');
+    test404('empty query causes 404', '?=', 'Bad request: "?=".');
+    test404('bad asset query', '?q=bob|bob',
+        'Could not find asset in build graph: bob|bob');
+    test404(
+        'bad path query',
+        '?q=bob/bob',
+        'Could not find asset for path "bob/bob". Tried:\n'
+        '- example|bob/bob\n'
+        '- example|web/bob/bob');
+    test404(
+        'valid path, 2nd try',
+        '?q=bob/initial.txt',
+        'Could not find asset for path "bob/initial.txt". Tried:\n'
+        '- example|bob/initial.txt\n'
+        '- example|web/bob/initial.txt');
+
+    void testSuccess(String testName, String path, String expectedId) {
+      test(testName, () async {
+        var response = await getResponse(path);
+
+        var output = await response.readAsString();
+        expect(response.statusCode, 200, reason: output);
+        var json = jsonDecode(output) as Map<String, dynamic>;
+
+        expect(json, containsPair('primary', containsPair('id', expectedId)));
+      });
+    }
+
+    testSuccess('valid path', '?q=web/initial.txt', 'example|web/initial.txt');
+    testSuccess('valid path, leading slash', '?q=/web/initial.txt',
+        'example|web/initial.txt');
+    testSuccess('valid path, assuming root', '?q=initial.txt',
+        'example|web/initial.txt');
+    testSuccess('valid path, assuming root, leading slash', '?q=/initial.txt',
```

Comment:
Should we treat a leading slash as forced to the root of the package?

Reply:
🤷‍♂️  – I'm thinking of a user copy-pasting from their browser.

Since we actually print the AssetId that is understood, I'm not sure it's needed...